rpc/rest/security/rest_security.go

```diff
@@ -0,0 +1,451 @@
+var signature = "secret"
+
+// Default expiration time for token/cookie
+var defaultExpTime = time.Hour
+
+// AuthenticatorAPI provides methods for handling permissions
+type AuthenticatorAPI interface {
+	// AddPermissionGroup adds new permission group. PG is defined by name and a set of URL keys. User with
+	// permission group enabled has access to that set of keys. PGs with duplicated names are skipped.
+	AddPermissionGroup(group ...*access.PermissionGroup)
+
+	// Validate serves as middleware used while registering new HTTP handler. For every request, token
+	// and permission group is validated.
+	Validate(provider http.HandlerFunc) http.HandlerFunc
+}
+
+// Settings defines fields required to instantiate authenticator
+type Settings struct {
+	// Authentication database, default implementation is used if not set
+	AuthStore AuthenticationDB
+	// List of registered users
+	Users []access.User
+	// Expiration time (token claim). If not set, default value of 1 hour will be used.
+	ExpTime time.Duration
+	// Cost value used to hash user passwords
+	Cost int
+	// Custom token signature. If not set, default value will be used.
+	Signature string
+}
+
+// Credentials struct represents simple user login input
+type credentials struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+}
+
+// Authenticator keeps information about users, permission groups and tokens and processes it
+type authenticator struct {
+	log logging.Logger
+
+	// Router instance automatically registers login/logout REST API handlers if authentication is enabled
+	router    *mux.Router
+	formatter *render.Render
+
+	// User database keeps all known users with permissions and hashed password. Users are loaded from
+	// HTTP config file
+	// TODO add option to register users
+	userDb AuthenticationDB
+	// Permission database is a map of name/permissions and bound URLs
+	groupDb map[string][]*access.PermissionGroup_Permissions
+
+	// Token claims
+	expTime time.Duration
+}
+
+// NewAuthenticator prepares new instance of authenticator.
+func NewAuthenticator(router *mux.Router, ctx *Settings, log logging.Logger) AuthenticatorAPI {
+	a := &authenticator{
+		router: router,
+		log:    log,
+		formatter: render.New(render.Options{
+			IndentJSON: true,
+		}),
+		groupDb: make(map[string][]*access.PermissionGroup_Permissions),
+		expTime: ctx.ExpTime,
+	}
+
+	// Authentication store
+	if ctx.AuthStore != nil {
+		a.userDb = ctx.AuthStore
+	} else {
+		a.userDb = CreateDefaultAuthDB()
+	}
+
+	// Set token signature
+	signature = ctx.Signature
+	if a.expTime == 0 {
+		a.expTime = defaultExpTime
+		a.log.Debugf("Token expiration time claim not set, defaulting to 1 hour")
+	}
+
+	// Process users in go routine, since hashing may take some time
+	go func() {
+		// Add admin-user, enabled by default, always has access to every URL
+		hash, err := bcrypt.GenerateFromPassword([]byte("ligato123"), ctx.Cost)
```

Comment:
Please don't store plaintext here, it's not safe and it's is really unnecessary to generate hash from a known password every time. Just store hashed password. Only generating password needs cost, comparing can be done with any cost since the cost is retrievable from the hashed password, hence: https://godoc.org/golang.org/x/crypto/bcrypt#Cost

Reply:
the default admin password need not be safe (after all it's written in the readme), in the future, it will be possible to change it and then the hash has to be calculated anyway. But I agree, it is not needed to do it every time during startup.